blocks/bricks/interfaces.py

```
@@ -127,45 +128,97 @@ class Initializable(RNGMixin, Brick):
+
+        super(Initializable, self).__init__(**kwargs)
+
+    def _validate_roles_schmes(self):
+        for role in self.parameter_roles:
+            if role not in self.initialization_schemes.keys():
+                found = False
+                for init_role in list(self.initialization_schemes.keys()):
+                    if isinstance(role, type(init_role)):
+                        self.initialization_schemes[role] = \
+                                        self.initialization_schemes[init_role]
+                        found = True
+                if not found:
+                    raise ValueError("There is no initialization_schemes"
+                                     " defined for {}".format(role))
 
     def _push_initialization_config(self):
+        self._collect_roles()
+        self._validate_roles_schmes()
         for child in self.children:
-            if isinstance(child, Initializable):
+            if (isinstance(child, Initializable) and
+                    hasattr(child, 'initialization_schemes')):
                 child.rng = self.rng
-                if self.weights_init:
-                    child.weights_init = self.weights_init
-        if hasattr(self, 'biases_init') and self.biases_init:
-            for child in self.children:
-                if (isinstance(child, Initializable) and
-                        hasattr(child, 'biases_init')):
-                    child.biases_init = self.biases_init
+                for role, scheme in self.initialization_schemes.items():
+                    child.initialization_schemes[role] = scheme
+
+    def _collect_roles(self):
+        if hasattr(self, 'parameters'):
+            for param in self.parameters:
+                for role in param.tag.roles:
+                    if role in self.initializable_roles:
+                        self.parameter_roles.update(set([role]))
+
+    def _initialize(self):
+        for param in self.parameters:
```

This is a dangerous code. First, if a parameter has more than one role, it is not clear at all how it should be initialized. I would rather raise an exception upon encountering such a situation. Also, we should choose the most specific initialization scheme (e.g. the one for `RECURRENT_WEIGHT` and not for just `WEIGHT` if the parameter is `RECURRENT_WEIGHT`).